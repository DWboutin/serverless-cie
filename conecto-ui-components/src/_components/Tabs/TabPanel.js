import React from 'react'
import classNames from 'classnames'

import PropTypes from 'prop-types'

const TabPanel = ({ className, children, inactive }) => {
  const classes = classNames({
    ['to-tabs-panel']: true,
    [`${className}`]: className,
    inactive: inactive,
  })

  return <div className={classes}>{children}</div>
}

TabPanel.defaultProps = {
  clickDisabled: false,
}

TabPanel.propTypes = {
  clickDisabled: PropTypes.bool.isRequired,
}

export default TabPanel
