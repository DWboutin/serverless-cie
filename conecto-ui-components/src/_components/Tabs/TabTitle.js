import React from 'react'
import classNames from 'classnames'

import PropTypes from 'prop-types'

const TabTitle = ({
  className,
  active,
  title,
  onClick,
  clickDisabled,
  index,
}) => {
  const classes = classNames({
    ['to-tabs-title']: true,
    [`${className}`]: className,
    'click-disabled': clickDisabled,
    active: active,
  })
  const compProp = {}

  if (!clickDisabled) {
    compProp.onClick = onClick
  }

  return (
    <li className={classes} data-index={index} {...compProp}>
      <span>{title}</span>
    </li>
  )
}

TabTitle.defaultProps = {
  clickDisabled: false,
}

TabTitle.propTypes = {
  className: PropTypes.string,
  active: PropTypes.bool,
  clickDisabled: PropTypes.bool,
  index: PropTypes.number.isRequired,
  title: PropTypes.string.isRequired,
  onClick: PropTypes.func,
}

export default TabTitle
