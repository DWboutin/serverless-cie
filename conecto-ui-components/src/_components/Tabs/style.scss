@import '../_base/colors';

.to-tabs {
  overflow: hidden;

  nav {
    ul {
      box-sizing: border-box;
      display: -webkit-box;
      display: -ms-flexbox;
      display: flex;
      -webkit-box-flex: 0;
      -ms-flex: 0 1 auto;
      flex: 0 1 auto;
      -webkit-box-orient: horizontal;
      -webkit-box-direction: normal;
      -ms-flex-direction: row;
      flex-direction: row;
      -ms-flex-wrap: wrap;
      flex-wrap: wrap;
      padding: 0;
      margin: 0;

      li {
        display: -webkit-box;
        display: -ms-flexbox;
        display: flex;
        box-sizing: border-box;
        -webkit-box-flex: 0;
        -ms-flex: 1;
        flex: 1;
        -webkit-box-orient: horizontal;
        -webkit-box-direction: normal;
        -ms-flex-direction: column;
        flex-direction: column;
        -ms-flex-wrap: wrap;
        flex-wrap: wrap;

        font-family: Sailec, Arial, Helvetica, sans-serif;
        font-size: sizeToRem(14);
        text-align: center;
        line-height: 1.5em;
        justify-content: center;
        height: 50px;
        cursor: pointer;
        color: $veryPaleBrown;
        border-bottom: 2px solid $veryPaleBrown;

        span {
          pointer-events: none;
        }

        &.active,
        &:hover {
          color: $darkBrown;
          border-bottom: 2px solid $darkBrown;
        }

        &.click-disabled {
          cursor: default;

          &:not(.active):hover {
            color: $veryPaleBrown;
            border-bottom: 2px solid $veryPaleBrown;
          }
        }
      }
    }
  }

  &-panels {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-box-orient: horizontal;
    -webkit-box-direction: normal;
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-transition: margin-left 300ms $tansitionEase;
    transition: margin-left 300ms $tansitionEase;
    will-change: margin-left;
    width: 100%;

    .to-tabs-panel {
      -webkit-flex-shrink: 0;
      -ms-flex-negative: 0;
      flex-shrink: 0;
      width: 100%;
      opacity: 1;
      -webkit-transition: opacity 150ms;
      transition: opacity 150ms;

      &.inactive {
        height: 0;
        padding: 0 !important;
        opacity: 0;
        pointer-events: none;
      }
    }
  }
}