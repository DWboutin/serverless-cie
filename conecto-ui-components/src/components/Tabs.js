import React from 'react'

import Grid, { Row, Col } from '../_components/Grid'
import Tabs, { TabPanel } from '../_components/Tabs'
import Button from '../_components/Button'

const Tags = () => (
  <Grid>
    <Row>
      <Col>
        <h2 style={{ opacity: 0.7 }}>Tabs</h2>
      </Col>
    </Row>
    <Row>
      <Col
        xs={{
          span: 12,
        }}
        sm={{
          span: 12,
        }}
        md={{
          span: 6,
        }}
        lg={{
          span: 6,
        }}
        xl={{
          span: 6,
        }}
      >
        <h3 style={{ opacity: 0.6 }}>Title at the top</h3>
        <Tabs
          titlePosition="top"
          focusedIndex={1}
          fixedItem={<Button>Test</Button>}
        >
          <TabPanel title="Test title 1" clickDisabled>
            Test panel 1
          </TabPanel>
          <TabPanel title="Test title 2" clickDisabled>
            Test panel 2
          </TabPanel>
          <TabPanel title="Test title 3" clickDisabled>
            Test panel 3
          </TabPanel>
        </Tabs>
      </Col>
      <Col
        xs={{
          span: 12,
        }}
        sm={{
          span: 12,
        }}
        md={{
          span: 6,
        }}
        lg={{
          span: 6,
        }}
        xl={{
          span: 6,
        }}
      >
        <h3 style={{ opacity: 0.6 }}>Title at the bottom</h3>
        <Tabs titlePosition="bottom">
          <TabPanel title="Test title 1">Test panel 1</TabPanel>
          <TabPanel title="Test title 2">Test panel 2</TabPanel>
          <TabPanel title="Test title 3">Test panel 3</TabPanel>
        </Tabs>
      </Col>
    </Row>
  </Grid>
)

Tags.defaultProps = {}

export default Tags
