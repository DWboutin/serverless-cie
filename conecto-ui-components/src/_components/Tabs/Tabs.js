import React from 'react'
import classNames from 'classnames'

import PropTypes from 'prop-types'

import TabTitle from './TabTitle'

class Tabs extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      selectedIndex: props.focusedIndex,
    }

    this.classesPrefix = 'to-tabs'

    this.handleTitleClick = this.handleTitleClick.bind(this)
  }

  componentDidUpdate(prevProps, prevState) {
    const { onChangeCallBack } = this.props

    if (prevProps.focusedIndex !== this.props.focusedIndex) {
      this.setState({
        selectedIndex: this.props.focusedIndex,
      })
    }

    if (
      onChangeCallBack &&
      this.state.selectedIndex !== prevState.selectedIndex
    ) {
      onChangeCallBack(this.state.selectedIndex)
    }
  }

  handleTitleClick(e) {
    this.setState({
      selectedIndex: parseInt(e.target.dataset.index),
    })
  }

  renderTitles() {
    const { selectedIndex } = this.state
    const { children } = this.props

    return (
      <nav>
        <ul>
          {React.Children.map(children, (child, index) => (
            <TabTitle
              title={child.props.title}
              onClick={this.handleTitleClick}
              index={index}
              active={index === selectedIndex}
              clickDisabled={child.props.clickDisabled}
            />
          ))}
        </ul>
      </nav>
    )
  }

  render() {
    const { selectedIndex } = this.state
    const { className, children, titlePosition, fixedItem } = this.props
    const classes = classNames({
      [this.classesPrefix]: true,
      [`title-pos-${titlePosition}`]: true,
      [`${className}`]: className,
    })

    return (
      <div className={classes}>
        {titlePosition === 'top' && this.renderTitles()}
        <div className="to-tabs-wrap">
          <div
            className="to-tabs-panels"
            style={{ marginLeft: `${selectedIndex * -100}%` }}
          >
            {React.Children.map(children, (child, index) =>
              React.cloneElement(child, {
                inactive: index !== this.state.selectedIndex,
              })
            )}
          </div>
          {fixedItem && (
            <div className="fixed-item">
              {React.cloneElement(fixedItem, {
                selectedIndex: this.state.selectedIndex,
              })}
            </div>
          )}
        </div>
        {titlePosition === 'bottom' && this.renderTitles()}
      </div>
    )
  }
}

Tabs.defaultProps = {
  titlePosition: 'top',
  focusedIndex: 0,
  fixedItem: null,
}

Tabs.propTypes = {
  children: PropTypes.any.isRequired,
  titlePosition: PropTypes.oneOf(['top', 'bottom']).isRequired,
  focusedIndex: PropTypes.number.isRequired,
  onChangeCallBack: PropTypes.func,
  fixedItem: PropTypes.any,
}

export default Tabs
