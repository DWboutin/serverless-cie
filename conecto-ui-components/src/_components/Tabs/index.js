import Tabs from './Tabs'
import TabPanel from './TabPanel'
import TabTitle from './TabTitle'

export { Tabs as default, TabPanel, TabTitle }
