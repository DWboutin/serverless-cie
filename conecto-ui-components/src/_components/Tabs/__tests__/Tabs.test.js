import React from 'react'
import { shallow, mount } from 'enzyme'

import Tabs, { TabPanel } from '../../Tabs'

const CustomComp = () => <div className="customComp">Test</div>

describe('Tabs', () => {
  it('should match snapshot with titles at the top by default', () => {
    const wrapper = shallow(
      <Tabs>
        <TabPanel title="Test title 1">Test panel 1</TabPanel>
        <TabPanel title="Test title 2">Test panel 2</TabPanel>
        <TabPanel title="Test title 3">Test panel 3</TabPanel>
      </Tabs>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should match snapshot with titles at the bottom', () => {
    const wrapper = shallow(
      <Tabs titlePosition="bottom">
        <TabPanel title="Test title 1">Test panel 1</TabPanel>
        <TabPanel title="Test title 2">Test panel 2</TabPanel>
        <TabPanel title="Test title 3">Test panel 3</TabPanel>
      </Tabs>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should match snapshot with a fixedItem', () => {
    const wrapper = shallow(
      <Tabs fixedItem={<CustomComp />}>
        <TabPanel title="Test title 1">Test panel 1</TabPanel>
        <TabPanel title="Test title 2">Test panel 2</TabPanel>
        <TabPanel title="Test title 3">Test panel 3</TabPanel>
      </Tabs>
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should have the prefix class', () => {
    const wrapper = shallow(
      <Tabs>
        <TabPanel title="Test title 1">Test panel 1</TabPanel>
        <TabPanel title="Test title 2">Test panel 2</TabPanel>
        <TabPanel title="Test title 3">Test panel 3</TabPanel>
      </Tabs>
    )

    expect(wrapper.instance().classesPrefix).toEqual('to-tabs')
    expect(wrapper.hasClass(wrapper.instance().classesPrefix)).toEqual(true)
  })

  it('should contain added classes', () => {
    const wrapper = shallow(
      <Tabs className="test-class-1 test-class-2">
        <TabPanel title="Test title 1">Test panel 1</TabPanel>
        <TabPanel title="Test title 2">Test panel 2</TabPanel>
        <TabPanel title="Test title 3">Test panel 3</TabPanel>
      </Tabs>
    )

    expect(wrapper.hasClass('test-class-1')).toEqual(true)
    expect(wrapper.hasClass('test-class-2')).toEqual(true)
  })

  it('should change selected index and margin-left on title clicks', () => {
    const wrapper = mount(
      <Tabs>
        <TabPanel title="Test title 1">Test panel 1</TabPanel>
        <TabPanel title="Test title 2">Test panel 2</TabPanel>
        <TabPanel title="Test title 3">Test panel 3</TabPanel>
      </Tabs>
    )

    wrapper
      .find('.to-tabs-title')
      .last()
      .simulate('click')

    expect(wrapper.state('selectedIndex')).toEqual(2)
    expect(wrapper.find('.to-tabs-panels').prop('style').marginLeft).toEqual(
      '-200%'
    )

    wrapper.unmount()
  })

  it('should change selected index when passed as focusedIndex prop', () => {
    const wrapper = mount(
      <Tabs>
        <TabPanel title="Test title 1">Test panel 1</TabPanel>
        <TabPanel title="Test title 2">Test panel 2</TabPanel>
        <TabPanel title="Test title 3">Test panel 3</TabPanel>
      </Tabs>
    )

    expect(wrapper.state('selectedIndex')).toEqual(0)

    wrapper.setProps({
      focusedIndex: 2,
    })

    expect(wrapper.state('selectedIndex')).toEqual(2)

    wrapper.unmount()
  })

  it('should clickDisabled prop put click-disabled class on component', () => {
    const wrapper = mount(
      <Tabs>
        <TabPanel title="Test title 1" clickDisabled>
          Test panel 1
        </TabPanel>
        <TabPanel title="Test title 2">Test panel 2</TabPanel>
        <TabPanel title="Test title 3">Test panel 3</TabPanel>
      </Tabs>
    )

    expect(
      wrapper
        .find('.to-tabs-title')
        .first()
        .hasClass('click-disabled')
    ).toEqual(true)

    wrapper.unmount()
  })

  it('should call the onChangeCallBack when changing state', () => {
    const onChangeCallBack = jest.fn()
    const wrapper = mount(
      <Tabs onChangeCallBack={onChangeCallBack}>
        <TabPanel title="Test title 1" clickDisabled>
          Test panel 1
        </TabPanel>
        <TabPanel title="Test title 2">Test panel 2</TabPanel>
        <TabPanel title="Test title 3">Test panel 3</TabPanel>
      </Tabs>
    )

    wrapper
      .find('.to-tabs-title')
      .last()
      .simulate('click')

    expect(onChangeCallBack).toBeCalled()
    expect(onChangeCallBack).toBeCalledWith(2)

    wrapper.unmount()
  })

  it('should the fixedItem contains the selectedIndex', () => {
    const wrapper = shallow(
      <Tabs fixedItem={<CustomComp />}>
        <TabPanel title="Test title 1">Test panel 1</TabPanel>
        <TabPanel title="Test title 2">Test panel 2</TabPanel>
        <TabPanel title="Test title 3">Test panel 3</TabPanel>
      </Tabs>
    )

    expect(wrapper.find('CustomComp').props().selectedIndex).toEqual(0)
  })
})
