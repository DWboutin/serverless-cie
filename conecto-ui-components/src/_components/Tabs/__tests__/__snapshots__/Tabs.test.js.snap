// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs should match snapshot with a fixedItem 1`] = `
<div
  className="to-tabs title-pos-top"
>
  <nav>
    <ul>
      <TabTitle
        active={true}
        clickDisabled={false}
        index={0}
        key=".0"
        onClick={[Function]}
        title="Test title 1"
      />
      <TabTitle
        active={false}
        clickDisabled={false}
        index={1}
        key=".1"
        onClick={[Function]}
        title="Test title 2"
      />
      <TabTitle
        active={false}
        clickDisabled={false}
        index={2}
        key=".2"
        onClick={[Function]}
        title="Test title 3"
      />
    </ul>
  </nav>
  <div
    className="to-tabs-wrap"
  >
    <div
      className="to-tabs-panels"
      style={
        Object {
          "marginLeft": "0%",
        }
      }
    >
      <TabPanel
        clickDisabled={false}
        inactive={false}
        key=".0"
        title="Test title 1"
      >
        Test panel 1
      </TabPanel>
      <TabPanel
        clickDisabled={false}
        inactive={true}
        key=".1"
        title="Test title 2"
      >
        Test panel 2
      </TabPanel>
      <TabPanel
        clickDisabled={false}
        inactive={true}
        key=".2"
        title="Test title 3"
      >
        Test panel 3
      </TabPanel>
    </div>
    <div
      className="fixed-item"
    >
      <CustomComp
        selectedIndex={0}
      />
    </div>
  </div>
</div>
`;

exports[`Tabs should match snapshot with titles at the bottom 1`] = `
<div
  className="to-tabs title-pos-bottom"
>
  <div
    className="to-tabs-wrap"
  >
    <div
      className="to-tabs-panels"
      style={
        Object {
          "marginLeft": "0%",
        }
      }
    >
      <TabPanel
        clickDisabled={false}
        inactive={false}
        key=".0"
        title="Test title 1"
      >
        Test panel 1
      </TabPanel>
      <TabPanel
        clickDisabled={false}
        inactive={true}
        key=".1"
        title="Test title 2"
      >
        Test panel 2
      </TabPanel>
      <TabPanel
        clickDisabled={false}
        inactive={true}
        key=".2"
        title="Test title 3"
      >
        Test panel 3
      </TabPanel>
    </div>
  </div>
  <nav>
    <ul>
      <TabTitle
        active={true}
        clickDisabled={false}
        index={0}
        key=".0"
        onClick={[Function]}
        title="Test title 1"
      />
      <TabTitle
        active={false}
        clickDisabled={false}
        index={1}
        key=".1"
        onClick={[Function]}
        title="Test title 2"
      />
      <TabTitle
        active={false}
        clickDisabled={false}
        index={2}
        key=".2"
        onClick={[Function]}
        title="Test title 3"
      />
    </ul>
  </nav>
</div>
`;

exports[`Tabs should match snapshot with titles at the top by default 1`] = `
<div
  className="to-tabs title-pos-top"
>
  <nav>
    <ul>
      <TabTitle
        active={true}
        clickDisabled={false}
        index={0}
        key=".0"
        onClick={[Function]}
        title="Test title 1"
      />
      <TabTitle
        active={false}
        clickDisabled={false}
        index={1}
        key=".1"
        onClick={[Function]}
        title="Test title 2"
      />
      <TabTitle
        active={false}
        clickDisabled={false}
        index={2}
        key=".2"
        onClick={[Function]}
        title="Test title 3"
      />
    </ul>
  </nav>
  <div
    className="to-tabs-wrap"
  >
    <div
      className="to-tabs-panels"
      style={
        Object {
          "marginLeft": "0%",
        }
      }
    >
      <TabPanel
        clickDisabled={false}
        inactive={false}
        key=".0"
        title="Test title 1"
      >
        Test panel 1
      </TabPanel>
      <TabPanel
        clickDisabled={false}
        inactive={true}
        key=".1"
        title="Test title 2"
      >
        Test panel 2
      </TabPanel>
      <TabPanel
        clickDisabled={false}
        inactive={true}
        key=".2"
        title="Test title 3"
      >
        Test panel 3
      </TabPanel>
    </div>
  </div>
</div>
`;
